File: Java/src/main/java/geo/model/Triangle.java
package geo.model;

/**
 * Triângulo definido por três pontos.
 *
 * <p>Espelha {@code triangle.h}. A área usa a Fórmula de Heron (a partir dos
 * três lados), idêntica à de {@code triangle.c}.</p>
 */
public final class Triangle extends Shape {

    private final Point p1;
    private final Point p2;
    private final Point p3;

    public Triangle(Point p1, Point p2, Point p3) {
        this.p1 = p1;
        this.p2 = p2;
        this.p3 = p3;
    }

    public Point getP1() {
        return p1;
    }

    public Point getP2() {
        return p2;
    }

    public Point getP3() {
        return p3;
    }

    @Override
    public double perimeter() {
        return p1.distanceTo(p2) + p2.distanceTo(p3) + p3.distanceTo(p1);
    }

    /** Área pela Fórmula de Heron (semi-perímetro). */
    @Override
    public double area() {
        double a = p1.distanceTo(p2);
        double b = p2.distanceTo(p3);
        double c = p3.distanceTo(p1);
        double s = (a + b + c) / 2.0;
        return Math.sqrt(s * (s - a) * (s - b) * (s - c));
    }

    @Override
    public String typeName() {
        return "TRIANGLE";
    }
}

File: C/src/triangle.c
#include "triangle.h"
#include <math.h>

Triangle triangle_create(Point p1, Point p2, Point p3) {
    Triangle t = {p1, p2, p3};
    return t;
}

double triangle_perimeter(Triangle t) {
    return point_distance(t.p1, t.p2) + 
           point_distance(t.p2, t.p3) + 
           point_distance(t.p3, t.p1);
}

double triangle_area(Triangle t) {
    double a = point_distance(t.p1, t.p2);
    double b = point_distance(t.p2, t.p3);
    double c = point_distance(t.p3, t.p1);
    double s = (a + b + c) / 2.0;
    return sqrt(s * (s - a) * (s - b) * (s - c));
}
